libshaderc_spvc/src/spvc.cc

```diff
@@ -0,0 +1,155 @@
+void shaderc_spvc_compile_options_release(
+    shaderc_spvc_compile_options_t options) {
+  delete options;
+}
+
+void shaderc_spvc_compile_options_set_target_env(
+    shaderc_spvc_compile_options_t options, shaderc_target_env target,
+    shaderc_env_version version) {
+  switch (target) {
+    case shaderc_target_env_opengl:
+    case shaderc_target_env_opengl_compat:
+      switch (version) {
+        case shaderc_env_version_opengl_4_5:
+          options->target_env = SPV_ENV_OPENGL_4_5;
+          break;
+        default:
+          break;
+      }
+      break;
+    case shaderc_target_env_vulkan:
+      switch (version) {
+        case shaderc_env_version_vulkan_1_0:
+          options->target_env = SPV_ENV_VULKAN_1_0;
+          break;
+        case shaderc_env_version_vulkan_1_1:
+          options->target_env = SPV_ENV_VULKAN_1_1;
+          break;
+        default:
+          break;
+      }
+      break;
+    default:
+      break;
+  }
+}
+
+void shaderc_spvc_compile_options_set_language_version(
+    shaderc_spvc_compile_options_t options, uint32_t version) {
+  options->glsl.version = version;
+}
+
+shaderc_spvc_compiler_t shaderc_spvc_compiler_initialize() {
+  return new (std::nothrow) shaderc_spvc_compiler;
+}
+
+void shaderc_spvc_compiler_release(shaderc_spvc_compiler_t compiler) {
+  delete compiler;
+}
+
+void consume_validation_message(shaderc_spvc_compilation_result* result,
+                                spv_message_level_t level, const char* src,
+                                const spv_position_t& pos,
+                                const char* message) {
+  result->validation_messages.append(message);
+  result->validation_messages.append("\n");
+}
+
+shaderc_spvc_compilation_result_t shaderc_spvc_compile_into_glsl(
+    const shaderc_spvc_compiler_t compiler, const uint32_t* source,
+    size_t source_len, shaderc_spvc_compile_options_t options) {
+  auto* result = new (std::nothrow) shaderc_spvc_compilation_result;
+  if (!result) return nullptr;
+  if (options->validate) {
+    spvtools::SpirvTools tools(options->target_env);
+    tools.SetMessageConsumer(std::bind(
+        consume_validation_message, result, std::placeholders::_1,
+        std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
+    if (!tools.Validate(source, source_len, spvtools::ValidatorOptions())) {
+      result->status = shaderc_compilation_status_validation_error;
+      return result;
+    }
+  }
+  TRY_IF_EXCEPTIONS_ENABLED {
+    spirv_cross::CompilerGLSL cross(source, source_len);
+    cross.set_common_options(options->glsl);
+    result->compiler_output = cross.compile();
+    // An exception during compiling would crash (if exceptions off) or jump to
+    // the catch block (if exceptions on) so if we're here we know the compile
+    // worked.
+    result->status = shaderc_compilation_status_success;
+  }
+  CATCH_IF_EXCEPTIONS_ENABLED(...) {
+    result->status = shaderc_compilation_status_compilation_error;
```

Comment:
I think this should also set the compiler output string to an error message saying SPIRV-Cross failed.  Otherwise we have no clue about the nature of the failure.